### ISSUE TITLE
java.io.EOFException: Attempt to read after eof.

### ISSUE THREAD

**Issue by Asura7969** [housepower/ClickHouse-Native-JDBC#83]
Hello!

I'm getting an error with insert batch query once in a while:
```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.worker.DriverWrapper$.main(DriverWrapper.scala:65)
	at org.apache.spark.deploy.worker.DriverWrapper.main(DriverWrapper.scala)
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 118 in stage 5221.0 failed 4 times, most recent failure: Lost task 118.3 in stage 5221.0 (TID 626642, 10.111.65.213, executor 35): java.sql.SQLException: Attempt to read after eof.
	at com.github.housepower.jdbc.connect.PhysicalConnection.receiveResponse(PhysicalConnection.java:88)
	at com.github.housepower.jdbc.connect.PhysicalConnection.receiveEndOfStream(PhysicalConnection.java:78)
	at com.github.housepower.jdbc.ClickHouseConnection.sendInsertRequest(ClickHouseConnection.java:138)
	at com.github.housepower.jdbc.statement.ClickHousePreparedInsertStatement.executeBatch(ClickHousePreparedInsertStatement.java:61)
	at com.hellobike.SoaToClickHouse$$anonfun$main$1$$anonfun$apply$1.apply(SoaToClickHouse.scala:107)
	at com.hellobike.SoaToClickHouse$$anonfun$main$1$$anonfun$apply$1.apply(SoaToClickHouse.scala:75)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:929)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:929)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2074)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:2074)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:109)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.EOFException: Attempt to read after eof.
	at com.github.housepower.jdbc.buffer.SocketBuffedReader.refill(SocketBuffedReader.java:67)
	at com.github.housepower.jdbc.buffer.SocketBuffedReader.readBinary(SocketBuffedReader.java:34)
	at com.github.housepower.jdbc.serializer.BinaryDeserializer.readVarInt(BinaryDeserializer.java:23)
	at com.github.housepower.jdbc.protocol.RequestOrResponse.readFrom(RequestOrResponse.java:32)
	at com.github.housepower.jdbc.connect.PhysicalConnection.receiveResponse(PhysicalConnection.java:86)
	... 15 more
```

**Comment by sundy-li** [comment]
Hi, which driver version did you use?
Could you show some code or show more detail description of usage?

**Comment by zhang2014** [comment]
And please provide the version of the clickhouse server version

**Comment by Asura7969** [comment]
```scala
soaStream.foreachRDD(rdd =>{

      rdd.foreachPartition(it => {

        InternalRedisClient.build(redisConfig)
        val jedis = InternalRedisClient.getPool.getResource
        val userTableSchema =
          InternalRedisClient.getTableSchema(jedis, redisETLkey).asScala.toList.zipWithIndex.sortBy(_._2)
        val cf = new ClickHouseConfig("jdbc:clickhouse://xxxx:9000/database", chConfig.value)
        val conn = ClickHouseConnection.createClickHouseConnection(cf)
        val insertSQL = ClickhouseUtils.initPrepareSQL(userTableSchema,clickhouseConfig.getProperty("tableName"))

        val pstmt = conn.prepareStatement(insertSQL)
        // bulkSize = 90000
        val bulkSize = clickhouseConfig.getProperty("insertBulkSize").toInt
        var length = 0
        val keys = userTableSchema.toMap.keys.map(_._1).toList.asJava
        while (it.hasNext) {
          length += 1
          val json = NewLogScan.scan(it.next().value(), keys)
          if (null != json){
            ClickhouseUtils.newInsert(json,pstmt,userTableSchema)
          }
          if (length >= bulkSize) {
            pstmt.executeBatch()
            length = 0
          }
        }
        pstmt.executeBatch()
        conn.close()
      })
    })
```
clickhouse version:19.4.1

**Comment by Asura7969** [comment]
```xml
<dependency>
      <groupId>com.github.housepower</groupId>
      <artifactId>clickhouse-native-jdbc</artifactId>
      <version>1.6-stable</version>
</dependency>
```

**Comment by Asura7969** [comment]
> Hi, which driver version did you use?
> Could you show some code or show more detail description of usage?

clickhouse version:19.4.1
```xml
<dependency>
      <groupId>com.github.housepower</groupId>
      <artifactId>clickhouse-native-jdbc</artifactId>
      <version>1.6-stable</version>
</dependency>
```

```scala
soaStream.foreachRDD(rdd =>{

      rdd.foreachPartition(it => {

        InternalRedisClient.build(redisConfig)
        val jedis = InternalRedisClient.getPool.getResource
        val userTableSchema =
          InternalRedisClient.getTableSchema(jedis, redisETLkey).asScala.toList.zipWithIndex.sortBy(_._2)
        val cf = new ClickHouseConfig("jdbc:clickhouse://xxxx:9000/database", chConfig.value)
        val conn = ClickHouseConnection.createClickHouseConnection(cf)
        val insertSQL = ClickhouseUtils.initPrepareSQL(userTableSchema,clickhouseConfig.getProperty("tableName"))

        val pstmt = conn.prepareStatement(insertSQL)
        // bulkSize = 90000
        val bulkSize = clickhouseConfig.getProperty("insertBulkSize").toInt
        var length = 0
        val keys = userTableSchema.toMap.keys.map(_._1).toList.asJava
        while (it.hasNext) {
          length += 1
          val json = NewLogScan.scan(it.next().value(), keys)
          if (null != json){
            ClickhouseUtils.newInsert(json,pstmt,userTableSchema)
          }
          if (length >= bulkSize) {
            pstmt.executeBatch()
            length = 0
          }
        }
        pstmt.executeBatch()
        conn.close()
      })
    })
```

**Comment by sundy-li** [comment]
For the second 
`pstmt.executeBatch()` 

maybe better to be 

```
if (length > 0) {
    pstmt.executeBatch()
}
```

I don't know if this is the problem,  just test for `v19.5.1`  and that's ok.

**Comment by sundy-li** [comment]
And I don't know the detail of `ClickhouseUtils.java`.
Could you run the sample codes in the `src/main/java/examples` directory?

**Comment by Asura7969** [comment]
> And I don't know the detail of `ClickhouseUtils.java`.
> Could you run the sample codes in the `src/main/java/examples` directory?

```scala
object ClickhouseUtils {

  /**
    * Clickhouse 连接信息
    */
  def getCHConf:Properties={
    val clickhouseConf = new Properties()
    clickhouseConf.put("user","xxxxx")
    clickhouseConf.put("password","xxxx")
    clickhouseConf.put("database","xxx")
    clickhouseConf
  }

  /**
    * 初始化 insert SQL
    */
  def initPrepareSQL(userTableSchema: immutable.Seq[((String, String), Int)],tableName:String): String = {
    val prepare = List.fill(userTableSchema.size)("?")

    val fields = userTableSchema.map(_._1._1).mkString(",")
    val sql = String.format(
      "INSERT INTO %s (%s) VALUES (%s)",
      tableName,
      fields,
      prepare.mkString(","))
    sql
  }

  /**
    * 校验是否是数值
    */
  def isIntByRegex(s : String) = {
    val pattern = """^(\d+)$""".r
    s match {
      case pattern(_*) => true
      case _ => false
    }
  }

  /**
    * 插入数据
    */
  def newInsert(json: JSONObject,
                statement: PreparedStatement,
                userTableSchema: immutable.Seq[((String, String), Int)]): Unit ={

    userTableSchema.foreach(each => {
      val field = each._1._1
      val fieldType = each._1._2
      val i = each._2

      val value = json.get(field)
      fieldType match {
        case "String" =>
          if (null == value) statement.setString(i + 1, "null")
          else statement.setString(i + 1, json.getString(field))

        case "DateTime" =>
          if (null != value) statement.setTimestamp(i + 1,new Timestamp(value.toString.toLong))
          else statement.setTimestamp(i + 1,new Timestamp(System.currentTimeMillis()))

        case "Date" =>
          if (null != value) statement.setDate(i + 1,new Date(value.toString.toLong))
          else statement.setDate(i + 1,new Date(System.currentTimeMillis()))

        case "Int8" | "UInt8" | "Int16" | "UInt16" | "Int32" =>
          if (null == value) statement.setInt(i + 1, 0)
          else statement.setInt(i + 1, json.getIntValue(field))

        case "UInt32" | "UInt64" | "Int64" =>
          if (null != value && isIntByRegex(value.toString.replace(".",""))){
            statement.setLong(i + 1, json.getLongValue(field))
          } else {
            statement.setLong(i + 1, 0)
          }

        case "Float32" =>
          if (null != value && isIntByRegex(value.toString.replace(".",""))){
            statement.setFloat(i + 1, json.getFloatValue(field))
          } else {
            statement.setFloat(i + 1, 0)
          }

        case "Float64" =>
          if (null != value && isIntByRegex(value.toString.replace(".",""))){
            statement.setDouble(i + 1, json.getFloatValue(field))
          } else {
            statement.setDouble(i + 1, 0)
          }

        // TODO: 暂时不考虑Array类型
        // case arrayPattern(_) =>
        // statement.setArray(i + 1, item.getAs[WrappedArray[AnyRef]](field))
        case _ =>
      }
    })
    statement.addBatch()
  }
}
```

**Comment by BernieJiangDL** [comment]
the same thing happened here. I've tried to use spark dataframe jdbc to write down some items, but got the "java.io.EOFException: Attempt to read after eof." 

The code I am using is like:
    df.write
      .mode(saveMode)
      .format("jdbc")
      .option("driver", "com.github.housepower.jdbc.ClickHouseDriver")
      .option("url", "jdbc:clickhouse:/xxxxxxxx")
      .save()

**Comment by Asura7969** [comment]
Later I used http:https://github.com/yandex/clickhouse-jdbc




------------------ 原始邮件 ------------------
发件人: "BernieJiangDL"<notifications@github.com>;
发送时间: 2019年8月30日(星期五) 上午10:58
收件人: "housepower/ClickHouse-Native-JDBC"<ClickHouse-Native-JDBC@noreply.github.com>;
抄送: "Asura"<1402357969@qq.com>;"Author"<author@noreply.github.com>;
主题: Re: [housepower/ClickHouse-Native-JDBC] java.io.EOFException: Attemptto read after eof. (#83)




the same thing happened here. I've tried to use spark dataframe jdbc to write down some items, but got the "java.io.EOFException: Attempt to read after eof."
 
The code I am using is like:
 df.write
 .mode(saveMode)
 .format("jdbc")
 .option("driver", "com.github.housepower.jdbc.ClickHouseDriver")
 .option("url", "jdbc:clickhouse:/xxxxxxxx")
 .save()

**Comment by BernieJiangDL** [comment]
Thanks for your response.

I've tried yandex/clickhouse-jdbc dependency at the very first time. However I ran into some issues :

 one of the dependency of yandex/clickhouse-jdbc named jackson-databind's version is banned because of security reasons.

How do you use it? Could you please show your demo and pom?

I really appriciate it.





Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by shizhenchina** [comment]
I have the same problem

**Comment by pan3793** [comment]
> I've tried yandex/clickhouse-jdbc dependency at the very first time. However I ran into some issues : one of the dependency of yandex/clickhouse-jdbc named jackson-databind's version is banned because of security reasons. How do you use it? Could you please show your demo and pom? I really appriciate it. 

Generally, use shade and relocate to avoiding such dependency conflicts.

If use `clickhouse-native-jdbc`, paste below code segment in `pom.xml`. See more details at #156 
```xml
<!-- (recommended) shaded version, available since 2.3-stable -->
<dependency>
    <groupId>com.github.housepower</groupId>
    <artifactId>clickhouse-native-jdbc-shaded</artifactId>
    <version>2.3-stable</version>
</dependency>
```
If use `yandex/clickhouse-jdbc`, paste below code segment in `pom.xml`
```xml
<dependency>
   <groupId>ru.yandex.clickhouse</groupId>
   <artifactId>clickhouse-jdbc</artifactId>
   <classifier>shaded</classifier>
   <version>${yandex-clickhouse-jdbc.version}</version>
   <exclusions>
      <exclusion>
         <groupId>*</groupId>
         <artifactId>*</artifactId>
      </exclusion>
   </exclusions>
</dependency>
```

**Comment by pan3793** [comment]
Close by https://github.com/housepower/ClickHouse-Native-JDBC/commit/e795d27acf0579e7c6589e353ec66995e669221f
